flutter_tools: Auto-generate ExportOptions.plist for manual iOS code signing

MohammedTarigg opened the pull request:
# flutter_tools: Auto-generate ExportOptions.plist for manual iOS code signing

## Summary

Enhance `flutter build ipa` to automatically generate a complete ExportOptions.plist for manual code signing configurations, eliminating the need for users to create and maintain one manually.

**Fixes:** #177853

## Problem

When using manual code signing (`CODE_SIGN_STYLE=Manual`) with `flutter build ipa` for Release/Profile builds, the archive succeeds but the export step fails with:

```
error: exportArchive: "Runner.app" requires a provisioning profile with the Push Notifications and Sign in with Apple features.
```

**Root cause:** Flutter generated an incomplete ExportOptions.plist that only included `method` and `uploadBitcode`. When using manual signing, `xcodebuild -exportArchive` requires:
- `teamID` (from project's `DEVELOPMENT_TEAM`)
- `signingStyle=manual`
- `provisioningProfiles` (mapping bundle ID to provisioning profile UUID)

Without these, Xcode cannot resolve the provisioning profile for export, even if the profile is correctly configured and contains the required entitlements.

## Solution

Enhanced `_createExportPlist()` in `BuildIOSArchiveCommand` to automatically generate a complete ExportOptions.plist when:
1. `CODE_SIGN_STYLE=Manual` is detected for the main app target
2. Build mode is Release or Profile (production builds only)
3. A valid provisioning profile can be located and parsed

The generated plist includes:
- `method` (app-store, ad-hoc, enterprise, etc. - respects user's `--export-method` flag)
- `teamID` (from `DEVELOPMENT_TEAM` build setting)
- `signingStyle=manual`
- `provisioningProfiles` mapping main app bundle ID to provisioning profile UUID

**Fallback behavior:**
- If profile lookup fails: silently fall back to simple plist (no regression)
- For Automatic signing: unchanged behavior
- For Debug builds: unchanged behavior (not App Store export)
- For multi-target apps (extensions): falls back to simple plist today (see TODO below)

## Changes

### Core Implementation
- Added `ProfileData` class to encapsulate provisioning profile info (UUID and name)
- Refactored profile handling into reusable static methods for testability
- Enhanced `_createExportPlist()` with manual signing detection and profile lookup
- Added `_findProvisioningProfileUuid()` to locate provisioning profiles by specifier
- Added `_parseProvisioningProfileInfo()` to decode provisioning profile data once
- Added comprehensive trace logging for debugging

### Testing
- Created `build_ipa_export_plist_test.dart` with 7 unit tests covering:
  - Manual signing with profile found → enhanced plist generated ✓
  - Automatic signing → simple plist (unchanged behavior) ✓
  - Debug builds → simple plist (unchanged behavior) ✓
  - Different export methods → respected in generated plist ✓
  - Profile lookup failures → fallback to simple plist ✓
  - Null codeSignStyle → handled gracefully ✓
- All existing flutter_tools tests continue to pass

## Documentation

Added extensive inline comments explaining:
- Enhancement conditions and fallback behavior
- Provisioning profile search strategy and error handling
- Performance and security considerations
- Future enhancements (multi-target support)

## Safety & Scope

### What This Fixes
- ✅ Manual signing export failures for apps with Push Notifications / Sign in with Apple
- ✅ Auto-generates correct plist for **any** entitlements covered by provisioning profile
- ✅ Unblocks users from manual `--export-options-plist` workaround
- ✅ Improves iOS build UX for enterprise teams doing manual signing

### What This Does NOT Change
- ✅ Automatic signing behavior (unchanged)
- ✅ Debug builds (unchanged)
- ✅ CLI interface (no new flags)
- ✅ Ad-hoc / enterprise distributions (unchanged if not using manual signing)

### Known Limitations (Future Enhancements)
- Currently only supports single-target apps (main app bundle ID only)
- TODO: Multi-target apps with extensions (notification service, widgets, watch, etc.) - each extension target bundle ID may need separate entry in provisioningProfiles dict
- TODO: Support for multiple provisioning profiles if app uses multiple signing identities

---

## Pre-merge Checklist

- [x] I have signed the CLA
- [x] I read the Contributor Guide and Tree Hygiene docs
- [x] I linked the issue this fixes (#177853) and explained the failure scenario
- [x] I added documentation and trace logs so developers understand what changed
- [x] I added comprehensive unit tests for the new behavior
- [x] All existing and new tests pass locally (`flutter test packages/flutter_tools`)
- [x] Code passes linting (`dart analyze`)

## Safety Notes

- [x] This code only runs for manual code signing in `flutter build ipa` for Release/Profile builds
- [x] Debug builds and automatic signing behavior are completely unchanged
- [x] If no provisioning profile UUID is found, we silently fall back to the old exportOptions.plist logic instead of failing
- [x] Added trace-level logging explaining success/fallback scenarios for debugging
- [x] No breaking changes - existing workarounds continue to work
- [x] Minimal blast radius - only affects manual signing export path

## Verification

Users can verify the fix by:
1. Creating an iOS app with `CODE_SIGN_STYLE=Manual`, `DEVELOPMENT_TEAM` set, and Push Notifications entitlements
2. Running `flutter build ipa --release --verbose`
3. Confirming the trace log shows "Generated ExportOptions.plist with teamID, signingStyle=manual, and provisioningProfiles"
4. Verifying the IPA is successfully created without needing `--export-options-plist`

---

## Related Issues
- #106612 - Support `flutter build ipa` with manual signing and provisioning profiles
- #113977 - Flutter build IPA with --export-options-plist not working

MohammedTarigg committed:
flutter_tools: generate enhanced ExportOptions.plist for manual signing

Automatically generate ExportOptions.plist with teamID, signingStyle, and
provisioningProfiles mapping when CODE_SIGN_STYLE=Manual is detected for
release or profile builds. This enables flutter build ipa to work correctly
for manual signing configurations without requiring users to manually provide
an ExportOptions.plist file.

The enhancement:
- Detects manual signing (CODE_SIGN_STYLE=Manual) in release/profile builds
- Extracts provisioning profile UUID from installed profiles
- Generates ExportOptions.plist with complete signing configuration
- Falls back to simple plist if profile cannot be found
- Only activates for release/profile builds (not debug)

Unit tests are included to verify the plist generation logic for:
- Automatic signing (unchanged behavior)
- Manual signing with profile found (enhanced plist)
- Manual signing with profile not found (fallback)
- Debug builds (simple plist even with manual signing)
- Profile builds (enhanced plist with manual signing)

Fixes https://github.com/flutter/flutter/issues/177853

MohammedTarigg committed:
flutter_tools: refactor provisioning profile handling in BuildIOSArchiveCommand

- Introduced ProfileData class to encapsulate provisioning profile UUID and name.
- Updated _findProvisioningProfileUuid to utilize ProfileData for improved clarity and efficiency.
- Enhanced _parseProvisioningProfileInfo to extract both UUID and name in a single operation.
- Added trace logging for better debugging when provisioning profiles cannot be found.
- Removed unused ProcessManager and PlistParser parameters from method signatures in BuildIOSArchiveCommand.

This refactor streamlines provisioning profile management and improves logging for manual signing scenarios.

MohammedTarigg committed:
flutter_tools: enhance ExportOptions.plist generation and improve provisioning profile handling

- Added detailed documentation for the ProfileData class and its usage in generating ExportOptions.plist.
- Implemented logic to auto-generate an enhanced ExportOptions.plist for manual signing scenarios, ensuring compatibility with Xcode's requirements.
- Improved logging for provisioning profile lookups, providing clearer trace messages for debugging.
- Expanded unit tests to cover various manual signing scenarios, including fallback behaviors and different build modes.

MohammedTarigg committed:
flutter_tools: document limitations and enhance logging for manual signing

MohammedTarigg committed:
refactor: apply reviewer suggestions for code style and async/await best practices

- Simplify FakeXcodeProjectInterpreter.getBuildSettings by defining defaults
  first, then spreading overrides (eliminates redundant ?? operators)
- Replace listSync() with await for (using list()) in _findProvisioningProfileUuid
  for better asynchronous code style and non-blocking file system operations

Addresses code review feedback from PR #177888

MohammedTarigg committed:
fix: add justification comments for skipped tests to satisfy tree hygiene

Add comments directly above skip parameters explaining why integration
tests are skipped. This satisfies Flutter's tree hygiene requirement
that all skipped tests must have justification comments.

Fixes CI failure in PR #177888

MohammedTarigg committed:
chore: retrigger CI (infra failure in Mac_arm64 tool_host_cross_arch_tests)

vashworth commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -532,7 +572,8 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
       status = globals.logger.startProgress('Building $exportMethodDisplayName IPA...');
       if (exportOptions == null) {
-        generatedExportPlist = _createExportPlist(exportMethod);
+        final BuildInfo buildInfo = await cachedBuildInfo;
```
Is this needed? Isn't `buildInfo` already defined above?

https://github.com/flutter/flutter/blob/7c71a1f7e35b45bfb7f476276f41ae920b91b09e/packages/flutter_tools/lib/src/commands/build_ios.dart#L515

vashworth commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -618,29 +659,349 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
-    // Create the plist to be passed into xcodebuild -exportOptionsPlist.
+  /// Test-only method to expose ExportOptions.plist generation logic for unit testing.
+  @visibleForTesting
+  static Future<File> createExportPlistForTesting({
```
I think introducing test-only code here can be avoided by moving `buildSettings` out of `_createExportPlist` and by making the method public.

For example:

```dart
    final Map<String, String>? buildSettings = await app.project.buildSettingsForBuildInfo(
      buildInfo,
    );
    generatedExportPlist = await createExportPlist(exportMethod, app, buildInfo, buildSettings);
 
 ...
 
 Future<File> createExportPlist(
    String exportMethod,
    BuildableIOSApp app,
    BuildInfo buildInfo,
    Map<String, String>? buildSettings,
  ) async {
```

Then in your test, you could do something like this:

```dart
    final BufferLogger logger = BufferLogger.test();
    final command = BuildIOSArchiveCommand(logger: logger, verboseHelp: false);
    final String exportMethod = 'app-store';

    command.createExportPlist(exportMethod, FakeBuildableIOSApp(), BuildInfo.release, {
      'CODE_SIGN_STYLE': 'MANUAL',
      ...
    });
```

https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#test-apis-belong-in-the-test-frameworks

MohammedTarigg replied:
Fixed - made createExportPlist public with buildSettings as a parameter, eliminating the need for a test-only method.

vashworth commented on packages/flutter_tools/test/commands.shard/hermetic/build_ipa_export_plist_test.dart:
```diff
@@ -0,0 +1,227 @@
+    test(
+      'full build ipa path generates plist for manual signing (integration)',
+      () {
+        // Integration test skipped - see justification comment below.
```
You don't need to leave an empty test in place for the integration test. This comment is better put in the PR description

vashworth commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -618,29 +659,349 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
+  }
+
+  /// Static version for testing.
+  static File _createManualSigningExportPlistStatic({
```
You shouldn't need a static version for testing, see comment above:

MohammedTarigg replied:
Fixed - removed the static `createExportPlistForTesting` method. The method is now public (`createExportPlist`) and tests call it directly with mock dependencies, following Flutter's testing best practices.

vashworth commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -618,29 +659,349 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
+    return _createSimpleExportPlistStatic(exportMethod: exportMethod, fileSystem: globals.fs);
+  }
+
+  /// Static version for testing.
```
You shouldn't need a static version for testing, see comment above:

MohammedTarigg replied:
Fixed - removed all static test-only methods (`createExportPlistForTesting`, `_createSimpleExportPlistStatic`, and `_createManualSigningExportPlistStatic`). These have been converted to instance methods. The main method is now public (`createExportPlist`) and tests call it directly with mock dependencies, following Flutter's testing best practices.

vashworth reviewed (commented):
Thanks for the contribution! Overall the concept sounds good to me. I left some comments on tweaking the implementation to make it more testable.

MohammedTarigg committed:
refactor: apply all reviewer feedback from PR #177888

- Made createExportPlist public with buildSettings parameter
- Removed test-only createExportPlistForTesting method
- Eliminated design documentation from class docs
- Cleaned up empty test placeholders
- Fixed logging to use logger instead of globals.logger
- Extracted provisioning profile directory lookup into reusable method

Tests now call the public method directly with mock dependencies.

MohammedTarigg commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -106,6 +125,27 @@ class _ImageAssetFileKey {
+/// - Support apps with extensions/widgets (multiple provisioning profiles).
+/// - Make plist values (`uploadBitcode`, `stripSwiftSymbols`) configurable.
+/// - Cache profile lookups within a single build invocation.
+///
```
removed the design decision documentation from the class doc, moved to PR description.

MohammedTarigg commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -618,29 +659,349 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
+  /// - Profile not found - logs trace with helpful context and returns null
+  ///
+  /// Returns the UUID of the matching profile, or null if not found or parsing fails.
+  Future<String?> _findProvisioningProfileUuid(String profileSpecifier) async {
```
Addressed - extracted the provisioning profile directory lookup into a reusable _getProvisioningProfileDirectory() method to reduce duplication.

MohammedTarigg committed:
refactor: remove static test-only methods, convert to instance methods

Removed static _createSimpleExportPlistStatic and _createManualSigningExportPlistStatic methods and converted them to instance methods _createSimpleExportPlist and _createManualSigningExportPlist. This follows Flutter's testing best practices by avoiding test-only static methods.

MohammedTarigg committed:
style: format _createSimpleExportPlist method signature

MohammedTarigg committed:
Refactor provisioning profile code to eliminate duplication

Addresses code review feedback on PR #177888 by extracting shared
provisioning profile logic into reusable components.

Changes:
- Made ProvisioningProfile class public and added uuid field
- Made parseProvisioningProfile() method public in XcodeCodeSigningSettings
- Extracted getProvisioningProfileDirectory() as a public function
- Updated build_ios.dart to use shared code from code_signing.dart
- Removed ProfileData class and duplicate parsing methods

This eliminates approximately 73 lines of duplicate code and establishes
a single source of truth for provisioning profile handling across
flutter_tools.

MohammedTarigg commented on PR_REPLY.md:
Sorry about that, this is my first open-source contribution.

vashworth commented on packages/flutter_tools/test/commands.shard/hermetic/build_ipa_export_plist_test.dart:
```diff
@@ -0,0 +1,93 @@
+import 'package:flutter_tools/src/build_info.dart';
+import 'package:flutter_tools/src/commands/build_ios.dart';
+
+import '../../general.shard/ios/core_devices_test.dart';
```
I believe this import is causing the analyzer to fail. Why is this imported?

MohammedTarigg committed:
Refactor XcodeCodeSigningSettings for testability and fix analyzer issues

- Refactor XcodeCodeSigningSettings instantiation to runCommand and pass it to createExportPlist for dependency injection.
- Remove core_devices_test.dart import in build_ipa_export_plist_test.dart to fix analyzer failure.
- Implement local FakeXcodeCodeSigningSettings, FakeBuildableIOSApp, and FakeIosProject in build_ipa_export_plist_test.dart using noSuchMethod for robust, self-contained testing.
- Remove stale TODO comments in build_ipa_test.dart and build_ipa_export_plist_test.dart.

vashworth commented:
@MohammedTarigg Sorry for the delay. Your PR looks good to me, but the CLA check is failing due to how you merged in changes. Can you rebase your PR to drop those commits?

MohammedTarigg commented:
@vashworth  Thanks for the review and heads up, I've rebased the branch to drop those merge commits. The CLA check should pass now. Let me know if there's anything else needed.

MohammedTarigg committed:
test: add test for enhanced ExportOptions.plist with manual signing

vashworth reviewed (approved):
LGTM!

MohammedTarigg committed:
Fix manual signing plist test by injecting FileSystemUtils

The test for generating enhanced ExportOptions.plist for manual signing
was failing because _findProvisioningProfileUuid used globals.fs and
globals.fsUtils instead of the injected FileSystem, making it impossible
to test with MemoryFileSystem.

Changes:
- Add optional fileSystemUtils parameter to createExportPlist
- Add required fileSystem and fileSystemUtils parameters to
  _findProvisioningProfileUuid
- Update test to create provisioning profiles directory structure in
  MemoryFileSystem and pass FakeFileSystemUtils with home directory path
- Add FakeFileSystemUtils class to test file

The changes are backwards compatible - existing code that doesn't pass
fileSystemUtils will continue to use globals.fsUtils as the default.

MohammedTarigg commented:
Thanks so much to @jmagman for catching the test failure and the kind words, and to @vashworth for the thorough review and guidance throughout.

vashworth commented:
Analyze is failing:

```
    info • The import of 'package:file/file.dart' is unnecessary because all of the used elements are also provided by the import of 'package:flutter_tools/src/base/file_system.dart' • packages/flutter_tools/test/commands.shard/hermetic/build_ipa_export_plist_test.dart:5:8 • unnecessary_import
    info • Specify the type of a local variable when the type is non-obvious • packages/flutter_tools/test/commands.shard/hermetic/build_ipa_export_plist_test.dart:110:7 • specify_nonobvious_local_variable_types
    info • Specify the type of a local variable when the type is non-obvious • packages/flutter_tools/test/commands.shard/hermetic/build_ipa_export_plist_test.dart:115:7 • specify_nonobvious_local_variable_types
    info • Specify the type of a local variable when the type is non-obvious • packages/flutter_tools/test/commands.shard/hermetic/build_ipa_export_plist_test.dart:128:7 • specify_nonobvious_local_variable_types
```

MohammedTarigg committed:
fix: Address lint warnings in build_ipa_export_plist_test.dart

okorohelijah reviewed (approved):
LGTM

okorohelijah committed:
Merge branch 'master' into improve-ios-export-options-manual-signing
